runtime/matchers.go

```diff
@@ -0,0 +1,40 @@
+package runtime
+
+import "strings"
+
+// EqualMatcher performs a case-sensitive equality match for request metadata keys
+func EqualMatcher(h string) ServeMuxOption {
+	return func(mux *ServeMux) {
+		mux.headerMatchers = append(mux.headerMatchers, func(key string) bool {
+			return key == h
+		})
+	}
+}
+
+// EqualFoldMatcher performs a case-insensitive equality match for request metadata keys
+func EqualFoldMatcher(h string) ServeMuxOption {
+	return func(mux *ServeMux) {
+		mux.headerMatchers = append(mux.headerMatchers, func(key string) bool {
+			return strings.EqualFold(key, h)
+		})
+	}
+}
+
+// PrefixMatcher performs a case-sensitive prefix match for request metadata keys
+func PrefixMatcher(h string) ServeMuxOption {
+	return func(mux *ServeMux) {
+		mux.headerMatchers = append(mux.headerMatchers, func(key string) bool {
+			return strings.HasPrefix(key, h)
+		})
+	}
+}
+
+// PrefixFoldMatcher performs a case-insensitive prefix match for request metadata keys
+func PrefixFoldMatcher(h string) ServeMuxOption {
+	h = strings.ToLower(h)
+	return func(mux *ServeMux) {
+		mux.headerMatchers = append(mux.headerMatchers, func(key string) bool {
+			return strings.HasPrefix(strings.ToLower(key), h)
+		})
```

Comment:
I think this could be handled without the ToLower by doing an EqualFold on the head of the key. For example, https://play.golang.org/p/lh_JIYnZ0s

I only mention this because EqualFold is a bit more involved than comparing two lowercase strings.